### ISSUE TITLE
Support raw Matlab script/functions embedded in master equations and mechanisms

### ISSUE THREAD

**Issue by jasonsherfey** [DynaSim/DynaSim#591]
Feature request: Support blocks of Matlab code embedded in equations and mechanism files.

Context: there are many things that one might want to do that can not be achieved using expressions on single lines. For instance, for loops, if-then statements, and subfunctions. 

Current status: only single line expressions/statements are supported at present.

Implementation: the user should be able to indicate a block of code is raw Matlab and then implement whatever they need the same way they would in a Matlab script. On the development side, this will involve adding parameter/variable/function namespaces and inserting the block of raw Matlab code directly into the solve_file. Functions declared in the block of raw Matlab code should be appended to the end of the solve_file; everything else should be inserted into the solve_file immediately after the proceeding expression (note: order matters in this case).

**Comment by kupiqu** [comment]
I've just realized that this is way more important than it seems and once implemented I would encourage its usage against current functions and conditions, which in fact I would suggest to remove.

I know this is **BOLD**, but the reasons are the following:

- Conditionals:

They are very limited (no elseif option) and faulty (an if-else conditional didn't work for me). Tried to fix both but it is very complicated thing to do, which made me realize the alternative suggested in this issue would be a much better approach.

- Functions:

The way current functions work is by substitution, which is very nice and smart, yet pretty much limited (because once substituted there is no room for any change) and complex/convoluted codewise speaking (hence also difficult to follow and prompt to bugs). For instance one cannot define a function and then change its values with conditionals. And although I didn't try,  I'm almost sure one cannot define the a function following an if-else conditional.

You may wonder why (oh why!) removing instead of simply deprecating:

I would in principle suggest deprecating instead of removing of course. The problem of this is that I anticipate the complexity of this implementation being much higher (much longer time to deliver and with lots of issues to fix) if we are going to keep current conditionals and functions, because it's not going to be the same to "simply" deal with variable names according to scope, than to handle current functions and conditionals as well.

Also note that if we do, then all limitations about current functions and conditionals will still be potentially affecting this new implementation as well, and we won't have an easy way to help users design their models differently (that is combining functions and conditionals with matlab snippets).

To conclude, I would suggest implementing this issue in the most simple and natural way, assuming that all mechanisms are in fact in matlab language unless they appear in `''` as we do in the main file. I think that for this to work, we would only need to:
- define ODEs as we do outside mechanisms, that is like this: `'ODE'`, instead of simply `ODE` (that would make it actually more coherent)
- apply proper scope to variables
- replace current functions/conditions by expressions just the way they are naturally used in Matlab
- move actual matlab functions within mechanisms to the end of the solve file

@jasonsherfey, what do you think?

**Comment by kupiqu** [comment]
Actually, perhaps the best option, in terms of modularity and clarity also for the solve file, and arguably the easiest implementation, might be to associate a matlab function to each mechanism and plug this to the end of the solve file.

This would allow forgetting about scopes inside the mechanisms, that is inside these functions.

Scope check would instead be applied to generate the arguments of the function but would not be applied inside the function itself.

**Comment by jasonsherfey** [comment]
Creating separate functions for each mechanism (which is how I originally implemented mechanisms; i.e., like NEURON .mod files) can dramatically increase simulation time in Matlab, probably due to overhead from Matlab passing variables between workspaces. It is much quicker to integrate the equations without external function calls. The same is true (even more so) for anonymous function calls, which is why I designed the integrator to evaluate expressions created after function substitution. 

I think eliminating conditional makes sense for the reasons you stated (limited, buggy). Functions are tricky because the way they are handled has a very significant impact on how long the simulation takes to run.

**Comment by kupiqu** [comment]
I can see it would be generally slower. 

However:

- DynaSim is not (all) about speed (other alternatives like C++ and Julia are better)
- I think using the coder would alleviate the overhead, although not sure how variables are passed in the mex file made by the coder (hopefully by reference)
- DynaSim would be much easier to maintain and develop, without all the complexity of DynaSim functions and conditionals
- Solve files would end up being much more human-friendly

But more importantly perhaps:

- It would allow combining functions and conditionals, which is now not possible

Also quite important for allowing flexibility while optimizing speed:

- Functions could have on-the-fly evaluation so if the anticipated result is zero (e.g., because for instance the conductance of a mechanism is zero), the body of the function does not need to be evaluated and the function can even return a 0 scalar value right away.

This is now not easily doable and these embedded mechanisms into expressions are evaluated even if conductances are set to 0.

How is that conductances are set to 0 and why is this important? Perhaps it's just me, but when "playing" with DynaSim models, it's not unusual (for me at least) to test adding mechanisms, so the complexity of the model increases rapidly, which has a strong impact on speed (and readibility). It may be that later on some of these mechanisms is not used but usually I don't remove the mechanisms but set them to "0" (on the course of a project one is not totally certain if that conductance will later change to a different than 0 value again). 

Having them into functions, would allow setting a conditional at the head of its body, so if the conductance is zero, instead of overhead, this would imply saving time.

**Comment by kupiqu** [comment]
Nice summary from here: 

https://www.mathworks.com/matlabcentral/answers/80001-speed-functions-and-overheads#answer_89745

> - When you call functions, you cannot avoid the overhead.
> - When you inline the code instead, the overhead vanishes, but other difficulties arise.
> - Tertium non datur.

**Comment by kupiqu** [comment]
I know that you have invested a lot of time implementing conditionals and functions as they are, for the benefit of speed. I really appreciate that effort.

I realized about all this complexity by the effort I needed to make to first understand and then improve current version of conditionals in DynaSim.

They now work with (as many) elseif conditions and a final else condition. Yet, I would encourage replacing them as well as DynaSim functions, to embrace the simplicity, flexibility and elegance of the language conditionals and functions (at the cost of speed; which can be improved using the coder as well as checking conductance values).

**Comment by jsherfey** [comment]
In my experience with development, most of the convoluted complexity that is frustrating pertains to the implementation of conditionals, not functions. Relatedly, most of the bugginess that I've experienced involves conditionals (or monitors, but that's another issue). I would be very happy to replace conditionals (and clean up the code) with the more powerful capabilities enabled by supporting raw Matlab. I also think it's very important to support conditional functions, which is currently not supported, as you discuss.

However, DynaSim was intended to enable model specification using conventional mathematical notation (i.e., like XPP), in addition to modular specification. That is the primary reason I think we should continue supporting functions specified as `f(x) = ...`. Secondarily, every model of mine will break if we eliminate that form of specification. 

Function substitution, which results in faster simulation but more obscure solve files, is already optional. There is an option `reduce_function_calls_flag` in `dsWriteDynaSimSolver` that can be set to `0` to produce a more readable solve file with functions defined/applied using anonymous function handles. There may be ways to clean up the way that functions are handled in DynaSim without eliminating support for `f(x) = ...`; I'll need to think about this more.

In summary:
- I'd like to eliminate conditionals, which will make the code easier to understand/develop,
- support raw Matlab for defining conditionals, functions, condition-dependent functions, etc,
- and to continue supporting functions specified using conventional math notation.

**Comment by jsherfey** [comment]
It's also worth noting that a design goal of mine has always been to make it as easy as possible to prototype complex models, which is why I've focused on both intuitive notation and modular specification. I'm strongly in favor of anything that furthers that goal.

**Comment by kupiqu** [comment]
Alright, that may be a nice compromise actually. If we eliminate conditionals and allow current DynaSim functions as well as Matlab functions (supporting all flexibility of a programming language), then it may be the best of the two worlds.

By the way I totally agree with:

> It's also worth noting that a design goal of mine has always been to make it as easy as possible to prototype complex models, which is why I've focused on both intuitive notation and modular specification. I'm strongly in favor of anything that furthers that goal.

But thought this was sort of orthogonal to this discussion and that we could manage to support this anyway with parsing.

**Comment by jsherfey** [comment]
I'm not sure what you are asking? What is "this"?

**Comment by kupiqu** [comment]
You mentioned about `intuitive notation and modular specification`

Intuitive notation is what we use for in the model script for describing the main ODE's

Not sure what you really mean by modular specification. If you mean mechanisms, then matlab functions are also modular. So I meant we could support parsing of mathematical expressions in mechanisms implemented otherwise as Matlab functions, similar to how the main script represents Matlab code with mathematical ODE's.

Not trying to convince you anymore about this, just saying that this seems orthogonal to whether or not use DynaSim functions as they are now.

**Comment by kupiqu** [comment]
In other words, I wasn't against mathematical notation. My proposal was to introduce them within single quotes as ODE's are introduced in the main model script and use parsing.

The rest of the content in the mechanism would then assumed to be Matlab code that we could directly used (I think this could simplify parsing, perhaps and allow for further flexibility?), jus as it is for matlab code in the main model script.
